### PR TITLE
Feat: add url arg support

### PR DESCRIPTION
**Pull request by zhiyu0729**
Signed-off-by: Zhiyu Wang <zhiyuwang.newbis@gmail.com>

Support passing command parameters through url

### CODE REVIEW

**Comment by calvin0327** [comment]
@zhiyuone thanks for you contribution, since we don't have E2E yet, I will test it manually.

File: charts/cloudtty/_crds/cloudshell.cloudtty.io_cloudshells.yaml [hunk 1]
```diff
@@ -111,6 +111,10 @@ spec:
               ttl:
                 format: int32
                 type: integer
+              urlArg:
+                description: 'UrlArg allow client to send command line arguments in
+                  URL (eg: http://localhost:7681?arg=foo&arg=bar)'
+                type: boolean
```
Comment (panpan0000) [inline charts/cloudtty/_crds/cloudshell.cloudtty.io_cloudshells.yaml:117]: type is not boolean

File: pkg/apis/cloudshell/v1alpha1/types.go [hunk 2]
```diff
@@ -66,6 +66,8 @@ type CloudShellSpec struct {
 	VirtualServiceConfig *VirtualServiceConfig `json:"virtualServiceConfig,omitempty"`
 	// PathPrefix specified a path prefix to access url, if not, the default path is used.
 	PathPrefix string `json:"pathPrefix,omitempty"`
+	// UrlArg allow client to send command line arguments in URL (eg: http://localhost:7681?arg=foo&arg=bar)
+	UrlArg bool `json:"urlArg,omitempty"`
```
Comment (panpan0000) [inline pkg/apis/cloudshell/v1alpha1/types.go:70]: not bool

File: controllers/cloudshell_controller.go [hunk 3]
```diff
@@ -242,7 +242,7 @@ func (c *CloudShellReconciler) CreateCloudShellJob(ctx context.Context, cloudshe
 
 	jobBytes, err := util.ParseTemplate(jobTmpl, struct {
 		Namespace, Name, Ownership, Command, Configmap string
-		Once                                           bool
+		Once, UrlArg                                   bool
```
Comment (panpan0000) [inline controllers/cloudshell_controller.go:245]: not bool

**Comment by panpan0000** [comment]
```
 .../cloudtty/_crds/cloudshell.cloudtty.io_cloudshells.yaml  | 2 +-
 controllers/cloudshell_controller.go                        | 6 +++---
 pkg/apis/cloudshell/v1alpha1/types.go                       | 2 +-
 3 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/charts/cloudtty/_crds/cloudshell.cloudtty.io_cloudshells.yaml b/charts/cloudtty/_crds/cloudshell.cloudtty.io_cloudshells.yaml
index 7efffae..465e9fd 100644
--- a/charts/cloudtty/_crds/cloudshell.cloudtty.io_cloudshells.yaml
+++ b/charts/cloudtty/_crds/cloudshell.cloudtty.io_cloudshells.yaml
@@ -114,7 +114,7 @@ spec:
               urlArg:
                 description: 'UrlArg allow client to send command line arguments in
                   URL (eg: http://localhost:7681?arg=foo&arg=bar)'
-                type: boolean
+                type: string
               virtualServiceConfig:
                 description: VirtualServiceConfig specifies some of the parameters
                   necessary to create the virtaulService.
diff --git a/controllers/cloudshell_controller.go b/controllers/cloudshell_controller.go
index b2479c7..2c3ddb0 100644
--- a/controllers/cloudshell_controller.go
+++ b/controllers/cloudshell_controller.go
@@ -241,9 +241,9 @@ func (c *CloudShellReconciler) CreateCloudShellJob(ctx context.Context, cloudshe
 	}

 	jobBytes, err := util.ParseTemplate(jobTmpl, struct {
-		Namespace, Name, Ownership, Command, Configmap string
-		Once, UrlArg                                   bool
-		Ttl                                            int32
+		Namespace, Name, Ownership, Command, Configmap, UrlArg string
+		Once                                                   bool
+		Ttl                                                    int32
 	}{
 		Namespace: cloudshell.Namespace,
 		Name:      fmt.Sprintf("cloudshell-%s", cloudshell.Name),
diff --git a/pkg/apis/cloudshell/v1alpha1/types.go b/pkg/apis/cloudshell/v1alpha1/types.go
index 94f19ce..2e0c41e 100644
--- a/pkg/apis/cloudshell/v1alpha1/types.go
+++ b/pkg/apis/cloudshell/v1alpha1/types.go
@@ -67,7 +67,7 @@ type CloudShellSpec struct {
 	// PathPrefix specified a path prefix to access url, if not, the default path is used.
 	PathPrefix string `json:"pathPrefix,omitempty"`
 	// UrlArg allow client to send command line arguments in URL (eg: http://localhost:7681?arg=foo&arg=bar)
-	UrlArg bool `json:"urlArg,omitempty"`
+	UrlArg string `json:"urlArg,omitempty"`
 }

 // VirtualServiceConfig specifies some of the parameters necessary to create the virtaulService.
--
```

**Comment by panpan0000** [comment]
By the way, Can you elaborate a bit why this parameter is needed?
And can you please also update the Document to reflect your change, please?

**Comment by zhiyu0729** [comment]
@panpan0000 UrlArg is a switch, which is origin from ttyd.
This parameter can be used to add parameters to the startup command in the browser.

such as
```
spec:
  commandAction: $0
  urlArg: true
```
we can use url args to exec command
http://localhost:52409/?arg=kubectl%20exec%20-it%20busybox%20--%20bash
it will enter busybox pod shell

Importantly, this is a native parameter of ttyd.

**Comment by panpan0000** [comment]
> 

Thank you very much for your explanation , @zhiyu0729 
But to achieve this goal, you can refer to below example which copy from the README.md:
```
apiVersion: cloudshell.cloudtty.io/v1alpha1
kind: CloudShell
metadata:
  name: cloudshell-sample
spec:
  configmapName: "my-kubeconfig"
  runAsUser: "root"
  commandAction: "kubectl -n kube-system logs -f kube-apiserver-cn-stack"
  once: false
```

you can just replace the `commandAction:` with "kubectl exec -it busybox --bash" to achieve the same goal.

**Comment by zhiyu0729** [comment]
> > 
> 
> Thank you very much for your explanation , @zhiyu0729 But to achieve this goal, you can refer to below example which copy from the README.md:
> 
> ```
> apiVersion: cloudshell.cloudtty.io/v1alpha1
> kind: CloudShell
> metadata:
>   name: cloudshell-sample
> spec:
>   configmapName: "my-kubeconfig"
>   runAsUser: "root"
>   commandAction: "kubectl -n kube-system logs -f kube-apiserver-cn-stack"
>   once: false
> ```
> 
> you can just replace the `commandAction:` with "kubectl exec -it busybox --bash" to achieve the same goal.

yes, commandAction is static, but url arg is dynamic.

**Comment by panpan0000** [comment]
@zhiyu0729 , thanks for reply.
The smallest unit for cloudtty is the "CloudShell CR".
So when the CR is created and applied, the  `kubectl command` has been fixed in `commandAction` field.
You have no opportunity  to modify the `kubectl command` during runtime unless modified the CR again.
That's why I don't understand your point .
Please forgive me if I missed something .

**Comment by zhiyu0729** [comment]
> @zhiyu0729 , thanks for reply. The smallest unit for cloudtty is the "CloudShell CR". So when the CR is created and applied, the `kubectl command` has been fixed in `commandAction` field. You have no opportunity to modify the `kubectl command` during runtime unless modified the CR again. That's why I don't understand your point . Please forgive me if I missed something .

@panpan0000 it's because `commandAction` can't modify unless modified the CR again, `urlarg` then useful.

in your case, the pod only look one pod logs, if i want to look other pod log, need to create another cr.
but if your use `urlarg`, just

```
apiVersion: cloudshell.cloudtty.io/v1alpha1
kind: CloudShell
spec:
  commandAction: kubectl -n kube-system logs -f $0
  urlArg: true
```

image
everything is very simple, and I don't need create countless pods to see my logs.

**Comment by panpan0000** [comment]
Now I get it.Thank you very much for your explanation.This future is awesome .very useful

**Pull request merged**